packages/market/contracts/MarketLogic.sol

```diff
@@ -0,0 +1,363 @@
+
+    /// @notice Function to create a supply
+	/// @dev will return an event with the event-Id
+	/// @param _propertiesDocumentHash document-hash with all the properties of the demand
+	/// @param _documentDBURL url-address of the demand
+	/// @param _assetId the asset Id
+    function createSupply(
+        string calldata _propertiesDocumentHash,
+        string calldata _documentDBURL,
+        uint _assetId
+    ) external {
+        require(_assetLogic.getAssetOwner(_assetId) == msg.sender, "wrong msg.sender");
+
+        allSupply.push(Supply({
+            propertiesDocumentHash: _propertiesDocumentHash,
+            documentDBURL: _documentDBURL,
+            assetId: _assetId
+        }));
+
+        uint supplyID = allSupply.length > 0 ? allSupply.length - 1 : 0;
+
+        emit createdNewSupply(msg.sender, supplyID);
+    }
+
+    /// @notice function to return the length of the allSupply-array in the database
+	/// @return length of the allDemansa-array
+    function getAllSupplyListLength() external view returns (uint) {
+        return allSupply.length;
+    }
+
+	/// @notice Function to create a agreement
+	/// @dev will return an event with the event-Id
+	/// @param _propertiesDocumentHash document-hash with all the properties of the demand
+	/// @param _documentDBURL url-address of the demand
+	/// @param _demandId the demand Id
+	/// @param _supplyId the supply Id
+    function createAgreement(
+        string calldata _propertiesDocumentHash,
+        string calldata _documentDBURL,
+        uint _demandId,
+        uint _supplyId
+    ) external {
+        Demand memory demand = allDemands[_demandId];
+        Supply memory supply = allSupply[_supplyId];
+
+        address supplyOwner = _assetLogic.getAssetOwner(supply.assetId);
+
+        require(
+            msg.sender == demand.demandOwner || msg.sender == supplyOwner,
+            "createAgreement: wrong owner when creating"
+        );
+
+        allAgreements.push(Agreement({
+            propertiesDocumentHash: _propertiesDocumentHash,
+            documentDBURL: _documentDBURL,
+            demandId: _demandId,
+            supplyId: _supplyId,
+            approvedBySupplyOwner: false,
+            approvedByDemandOwner: false
+        }));
+
+        uint agreementId = allAgreements.length > 0 ? allAgreements.length - 1 : 0;
+
+        if (msg.sender == demand.demandOwner){
+            approveAgreementDemand(agreementId);
+        }
+
+        if (msg.sender == supplyOwner){
+            approveAgreementSupply(agreementId);
+        }
+
+        emit LogAgreementCreated(agreementId, _demandId, _supplyId);
+    }
+
+    /// @notice Matches a certificate to a demand from the agreement
+	/// @dev will return an event with the event-Id
+	/// @param _demandId index of the demand in the allDemands-array
+    /// @param _certificateId ID of the certificate
+    function fillAgreement(uint _demandId, uint _certificateId) external onlyRole(RoleManagement.Role.Matcher) {
```

Comment:
`fillAgreement` implementation is now the same as `fillDemand`. The idea of filling the agreement is (was?) based on the fact that certificates, in that case, were not set to be  `onSale` and thus `fillAgreement` was using `transferFrom` instead of `buyCertificateFor`.

Reply:
Thanks for the warning.